src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -1688,4 +1760,162 @@ bool RaftReplDev::is_resync_mode() {
+
+    const auto chunk_id = m_no_space_left_error_info.chunk_id;
+    const auto lsn = m_no_space_left_error_info.lsn;
+
+    HS_LOG_ASSERT(-1 != chunk_id, "chunk_id should be valid when handling no_space_left error, chunk_id={}, lsn={}",
+                  chunk_id, lsn);
+
+    RD_LOGD("start handling no_space_left error for chunk {} , lsn {}", chunk_id, lsn);
+
+    enter_emergency();
+
+    // 1. wait for all the pending reqs to be initialized and refuse later coming new requests
+    RD_LOGD("enter emergency state, waiting for all the pending req to be initialized");
+    while (true) {
+        uint64_t pending_req_num = get_pending_init_req_num();
+        if (pending_req_num) {
+            RD_LOGD("wait for {} pending create_req requests to be completed", pending_req_num);
+            std::this_thread::sleep_for(std::chrono::seconds(1));
+        } else
+            break;
+    }
+
+    // ensure the latest durable commit lsn is flushed to superblk before handling no_space_left
+    hs()->cp_mgr().trigger_cp_flush(true).wait();
+
+    // when reaching here, no pending req is being initailized, and no new req can be initailized (reqs from both data
+    // channel and raft channel will all be refused) from now on. before handling no_space_left error, we definitely
+    // need this precondition. for nuobject case, if new blks is allocated from data channel in the move_from
+    // chunk(emergent chunk) after the chunk is reset(be gced, so can allocate new blk), it will be dangerous , since
+    // now the emergent chunk will be returned to the chunkselector as a candidated chunk for new shard.
+
+    // 2. clean all the necessary rreqs in memory
+
+    // those reqs generated by data_channel might have successfully alloc blk in the emergent chunk, and thus the
+    // rreq has a state of BLK_ALLOCATED. if we do not clear them, after handling `no_space_left` error, when the
+    // log comes in raft channel, it will find the rreq already exists and block is already allocated, then it will
+    // skip allocating a new block for the log entry, which will lead to a data loss in the end. so we need to clear
+    // necessary the rreqs(which has allocated blk in the emergent chunk) in memory before handling `no_space_left`
+    // error.
+    RD_LOGD("start cleaning all the in-memory rreqs, which has allocated blk on the emergent chunk={} before handling "
+            "no_space_left error",
+            chunk_id);
+    std::vector< folly::Future< folly::Unit > > futs;
+    for (auto& [key, rreq] : m_repl_key_req_map) {
+        if (rreq->has_state(repl_req_state_t::BLK_ALLOCATED)) {
+            auto blkid = rreq->local_blkid();
+            if (chunk_id == blkid.chunk_num()) {
+                // only clean the rreqs which has allocated blks on the emergent chunk
+                futs.emplace_back(
+                    std::move(data_service().async_free_blk(blkid).thenValue([this, &blkid, &key](auto&& err) {
```

Comment:
it makes sence to remove those key from `m_repl_key_req_map` as they contains legacy local_blk_id.
however I dont get the necessity of async_free_blk

Reply:
for nuobject , we don`t need the necessity of async_free_blk since we will reset the chunk. but for other case(for example , nublox), which not use append_blk_allocator, not sure if they need this. so I add it here to make sure.